### PR TITLE
Extension framework

### PR DESCRIPTION
**Pull request by KennethNielsen**
I have made a simple framework for plugins, that will allow us to put all that code that is specific to just a single music service (including its xml requests) each into its own plugin, so that we don't clutter up the main class to much and make it too big.

You essentially create a plugin by creating a file inside the plugins dir that contains as its primary class one that is named in the same way is the file only in CamelCase, e.g. class Wimp() inside wimp.py. this class should take an instance of soco as its fist argument and a dictionary of additional arguments as is second, and that's it. Then the plugin can be loaded as soco.set_plugin('Wimp').

The following code illustrates how the plugins can be used:

```
import soco
import time

g = soco.SoCo('192.168.0.110')
plugin = g.get_plugin()
print 'Before we load a pluing it just returns us {0}'.format(plugin)
g.set_plugin('ExamplePlugin', {'user_name': 'Scooby Doo'})
plugin = g.get_plugin()
print 'Afterwards it returns us a reference to the plugin instance {0}'\
    .format(plugin)
# We can acces it through the variable it is stored in in SoCo
print 'Acessing though SoCo'
g.plugin.music_plugin_stop()
time.sleep(1)
g.plugin.music_plugin_play()
time.sleep(5)

# You can also get the music plugin and access its functions directly
print 'Accessing it directly'
plugin = g.get_plugin()
plugin.music_plugin_stop()
time.sleep(1)
plugin.music_plugin_play()
```

Let me know what you guys think. If this is usable I will of course write documentation on how to implement a plugin.

### CODE REVIEW

**Comment by stefankoegl** [comment]
I had a look at the interface you're describing (not yet at the actual implementation), but it does not seem to be very flexible.

As `SoCo` takes care of loading and instantiating the plugin, it depends on files being located in a specific directory. This might not be feasible, especially when SoCo is installed as a package. It also limits the user to one plugin at a time.

Why not design the plugin system the other way round, so that we provide an interface, such as 

``` python
class SoCoExtension(object):

    def __init__(self, soco):
        self.soco = soco

    # common methods go here
```

from which other plugins are derived as

``` python
class ExamplePlugin(SoCoExtension):

    def __init__(self, soco, other_param):
        super(ExamplePlugin, self).__init__(soco)
        pass

    # plugin-specific methods go here
```

When using a plugin, the user loads and instantiates the plugins (which could by located anywhere in `sys.path`) by himself

``` python

from example import ExamplePlugin

myplugin = ExamplePlugin(soco, 42)
myplugin.use()
```

I do not see a reason for storing a reference to the plugin in `SoCo`, so I'd rather let the user have the flexibility of handling plugins as it fits into his architecture.

What do you think?

**Comment by KennethNielsen** [comment]
OK, so even though the project has moved, I guess we can still finish this discussion here.

Just to be clear. The proposed solution doesn't restrict the location of plugins to a single directory. The plugins subdir is simply added to sys.path, and then loaded like anything else. So it would be possible to develop your own plugin from any folder you have added to sys.path.

That being said, it is restrictive, but it is so on purpose. The reason I wrote it like this is that I think it fits well with the way the system is used anyway. Something that I don't think will change very much with client implementation architecture.
1) Using a plugin is associated with a specific controller, either a single one or the master of a group, so linking the plugin to the general controller seems to fit. This also might potentially solve some of the problems mentioned about ungrouping if you call certain functions on slaves. This way you only need to keep of you soco units and which one of the is the master, then the rest will take care of itself.
2) With a controller you can also only use one plugin at a time, which makes the choice of limiting the container to one active plugin fit nicely. It would off course be trivially extendable to multiple plugins contained within soco, but I am not sure I see the point. One thing though that might trip up the idea of having just one, is if the loading of the plugin actually takes time, like if we imagine reading in a cached copy of the music tree from a specific service or something, or maybe as in the Pandora case to rely on internet communication for source info. This may be a deal breaker, I'm not sure.

So to sum up, I would make the case that yes it is restrictive, but I think that this will actually bring some advantage to most client applications. So the question is if you think the potential disadvantage for a client where it doesn't fit is large enough to be the determining factor. Anyway, I'm certainly open, and for sure your approach will also work fine. Let me know what you think.

**Comment by stefankoegl** [comment]
I see your points, but I still largely disagree with the interface.
- Adding your path to sys.path just to load a plugin is ugly, and might fail in many ways
- The interface I proposed also links the plugin to a specific controller, so I'm not quite sure if I understand your 1)
- I don't think you're always limited to just one plugin. Think of eg a last.fm plugin running in the background to extract currently playing songs for scrobbling
- Loading an extension by name adds a number of constraints (eg you must use a top-level module --  exactly one module per extension -- which is reloaded every time you switch extensions, potentially involving additional IO)
- Accessing the plugin through `mysoco.plugin` makes it totally un-obvious which plugin you're accessing. This is especially bad because most plugins will have different methods
- Your interface won't work in a multi-threaded environment (think GUI application with UI and worker threads), because

``` python
g.set_plugin('ExamplePlugin', {'user_name': 'Scooby Doo'})
# you can not be sure that you'll retrieve the plugin you just loaded, 
# because another thread could have loaded another one
plugin = g.get_plugin()
plugin.foo() # might raise an AttributeError
```

Whereas the following is imho much more explicit (and [Explicit is better than implicit](http://www.python.org/dev/peps/pep-0020/)) and thread-safe

``` python
from myplugins.example import ExamplePlugin
plugin = ExamplePlugin(g, user_name='Scooby Doo')
plugin.foo()
```

So while I understand your arguments why it might not be _so bad_ to have those restrictions, I don't see the point why we should have them. All they add are constraints and complexitiy, imho.

**Comment by KennethNielsen** [comment]
Hallo Stefan

About adding a path to sys.path. While it may not be pretty, can you explain to me one of the _many ways_ in which that can fail, considering that we through setup.py completely control the relative locations of the files to one another?

My point about 1 was to have 1 object per sonos unit, instead of (1 + the number of plugins) per sonos unit to keep track of.

In any case. Weighing pros and cons, and excluding the points above of which I am still uncertain, I think your solution still comes out clearly ahead, so let's continue with that. Will you write it up?

**Comment by stefankoegl** [comment]
I've created a pull request with my proposed implementation in SoCo/SoCo#2.

Imho the main problem with all those `sys.path` tricks is that it's hard to ensure that you're actually importing from the intended path. Other modules (and even threads) might do their own modifications. All this becomes especially complex when
- the `sys.path` modifications are not done at program startup, because when other modules do the same, the order of paths in `sys.path` depends on the order of instantiation/initialization.
- when there are restrictions such as `modulename == classname.lower()` which can easily cause naming conflicts with other modules

Anyway, I'd suggest to continue the discussion in SoCo/SoCo#2 due to the repository move.

**Comment by KennethNielsen** [comment]
Ahh yes. I see your point.

I look forward to have a look at your implementation. We talk more in the other thread.
